[Concurrency] Handle partial application of actor-isolated methods.

DougGregor opened the pull request:
Implement appropriate isolation checking for partial application of actor-isolated methods.

DougGregor committed:
[Concurrency] Prohibit partial application of actor-isolated methods.

DougGregor committed:
[Concurrency] Handle non-escaping partial application.

DougGregor commented:
@swift-ci please smoke test

theblixguy commented on lib/Sema/TypeCheckConcurrency.cpp:
```diff
@@ -338,6 +338,69 @@ class IsolationRestriction {
+    ApplyExpr *apply, Expr *parent) {
+  // Check for a call to the outer closure in the thunk.
+  auto outerAutoclosure = dyn_cast<AutoClosureExpr>(apply->getFn());
+  if (!outerAutoclosure ||
```
I think you can use `getUnwrappedCurryThunkExpr()` on `outerAutoclosure` to dig out the `DeclRefExpr` of the function.

DougGregor replied:
I didn't know this function existed! I'll figure out how to use it, thank you.

DougGregor replied:
Heh, good news. Doing do improved test coverage for `getUnwrappedCurryThunkExpr()`, so now both my new client and all existing clients handle more curry think ASTs properly.

slavapestov commented:
Might be worth adding some SILGen tests for the valid cases.

DougGregor commented:
> Might be worth adding some SILGen tests for the valid cases.

We don't have any SILGen support for asynchronous functions yet, so we can't write these tests meaningfully.

DougGregor commented:
Windows issue is unrelated:
```
mt.exe : general error c101008d: Failed to write the updated manifest to the resource of file "dispatch_timer_short.exe". Access is denied.
```

DougGregor committed:
Clean up AutoClosureExpr::getUnwrappedCurryThunkExpr().

Replace a hand-rolled variant of the subject function used for actor
isolation checking with a call to it. This exposed some limitations of
the function, which would assert() when dealing with partial
applications that involved optionals. Fix those as well.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
Source compatibility suite testing is still being adjusted for the master -> main transition.
